relay-replays/src/recording.rs

```diff
@@ -0,0 +1,662 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::io::{self, BufRead, Read, Write};
+
+use relay_general::pii::{DataScrubbingConfig, PiiProcessor};
+use relay_general::processor::{FieldAttrs, Pii, ProcessingState, Processor, ValueType};
+use relay_general::types::{Meta, ProcessingAction};
+
+use flate2::read::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use serde::de::Error as DError;
+use serde::{Deserialize, Serialize};
+use serde_json::{Error, Value};
+
+pub fn process_recording(bytes: &[u8]) -> Result<Vec<u8>, RecordingParseError> {
+    // Split recording headers and body.
+    let cursor = io::Cursor::new(bytes);
+    let mut split_iter = cursor
+        .split(b'\n')
+        .map(|r| r.map_err(|e| RecordingParseError::Message(e.to_string())));
+    let header = split_iter.next().ok_or_else(|| {
+        RecordingParseError::Message("no headers found. was data provided?".to_string())
+    })??;
+    let body = split_iter.next().ok_or_else(|| {
+        RecordingParseError::Message("no data found. are the headers missing?".to_string())
+    })??;
+
+    // Deserialization.
+    let mut events = loads(body)?;
+
+    // Processing.
+    strip_pii(&mut events).map_err(RecordingParseError::ProcessingAction)?;
+
+    // Serialization.
+    let out_bytes = dumps(events)?;
+    Ok([header, vec![b'\n'], out_bytes].concat())
+}
+
+fn loads(zipped_input: Vec<u8>) -> Result<Vec<Event>, RecordingParseError> {
+    let mut decoder = ZlibDecoder::new(zipped_input.as_slice());
```

Comment:
```suggestion
fn loads(zipped_input: &[u8]) -> Result<Vec<Event>, RecordingParseError> {
    let mut decoder = ZlibDecoder::new(zipped_input);
```

I think we can make this work without allocating a vector for `body`. `bytes.split(|b| b == &b'\n')` should give you an iterator of `&[u8]`, which you could pass directly to this function.

`header` would then also be a `&[u8]` so you'd have to pass `header.into()` into `concat()`.